go/acct/radius.go

```diff
@@ -0,0 +1,334 @@
+package main
+
+import (
+	"context"
+	"crypto/md5"
+	"database/sql"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/inverse-inc/go-radius"
+	"github.com/inverse-inc/go-radius/dictionary"
+	"github.com/inverse-inc/go-radius/rfc2865"
+	"github.com/inverse-inc/go-radius/rfc2866"
+	"github.com/inverse-inc/go-radius/rfc2869"
+	"github.com/inverse-inc/packetfence/go/db"
+	"github.com/inverse-inc/packetfence/go/jsonrpc2"
+	"github.com/inverse-inc/packetfence/go/log"
+	"github.com/inverse-inc/packetfence/go/mac"
+)
+
+var radiusDictionary *dictionary.Dictionary
+
+func (h *PfAcct) ServeRADIUS(w radius.ResponseWriter, r *radius.Request) {
+	if r.Code == radius.CodeAccountingRequest {
+		h.HandleAccountingRequest(w, r)
+		return
+	}
+}
+
+func (h *PfAcct) HandleAccountingRequest(w radius.ResponseWriter, r *radius.Request) {
+	outPacket := r.Response(radius.CodeAccountingResponse)
+	rfc2865.ReplyMessage_SetString(outPacket, "Accounting ok")
+	w.Write(outPacket)
+	ctx := r.Context()
+	iSwitchInfo := ctx.Value(switchInfoKey)
+	if iSwitchInfo == nil {
+		panic("SwitchInfo: not found")
+	}
+	switchInfo := iSwitchInfo.(*SwitchInfo)
+	go h.handleAccountingRequest(r, switchInfo)
+}
+
+func (h *PfAcct) handleAccountingRequest(r *radius.Request, switchInfo *SwitchInfo) {
+	//Acct-Output-Gigawords
+
+	callingStation := rfc2865.CallingStationID_GetString(r.Packet)
+	mac, _ := mac.NewFromString(callingStation)
+	in_bytes := int64(rfc2866.AcctInputOctets_Get(r.Packet))
+	out_bytes := int64(rfc2866.AcctOutputOctets_Get(r.Packet))
+	giga_in_bytes := int64(rfc2869.AcctInputGigawords_Get(r.Packet))
+	giga_out_bytes := int64(rfc2869.AcctOutputGigawords_Get(r.Packet))
+	out_bytes += giga_out_bytes << 32
+	in_bytes += giga_in_bytes << 32
+	timestamp := rfc2869.EventTimestamp_Get(r.Packet)
+	timestamp = timestamp.Truncate(h.TimeDuration)
+	h.InsertBandwidthAccounting(
+		switchInfo.TenantId,
+		mac.String(),
+		h.accountingUniqueSessionId(r),
+		timestamp,
+		in_bytes,
+		out_bytes,
+	)
+	h.sendRadiusAccounting(r)
+}
+
+func (h *PfAcct) accountingUniqueSessionId(r *radius.Request) string {
+	username := rfc2865.UserName_Get(r.Packet)
+	callingStation := rfc2865.CallingStationID_Get(r.Packet)
+	acctSessionId := rfc2866.AcctSessionID_Get(r.Packet)
+	hash := md5.New()
+	hash.Write(username)
+	hash.Write([]byte{','})
+	hash.Write(callingStation)
+	hash.Write([]byte{','})
+	hash.Write(acctSessionId)
+	sum := hash.Sum(nil)
+	return hex.EncodeToString(sum)
+}
+
+func (h *PfAcct) sendRadiusAccounting(r *radius.Request) {
+	attr := packetToMap(r.Context(), r.Packet)
+	attr["PF_HEADERS"] = map[string]string{
+		"X-FreeRADIUS-Server":  "packetfence",
+		"X-FreeRADIUS-Section": "accounting",
+	}
+	client := jsonrpc2.NewAAAClientFromConfig(r.Context())
```

Comment:
This creates tons of useless calls to pfconfig, you should create the client once and reuse it

Reply:
Fixed